research/minigo/minigo.py

```diff
@@ -0,0 +1,406 @@
+    estimator_working_dir: tf.estimator working directory.
+    params: Hyperparameters.
+  """
+  model_num, model_name = model_utils.get_latest_model(models_dir)
+
+  # Get the holdout game data
+  nums_names = model_utils.get_models(models_dir)
+  models = [num_name for num_name in nums_names if num_name[0] < model_num]
+
+  holdout_dirs = [os.path.join(holdout_dir, pair[1])
+                  for pair in models[-params.holdout_generation:]]
+  checkpoint_name = os.path.join(models_dir, model_name)
+
+  tf_records = []
+  with utils.logged_timer('Building lists of holdout files'):
+    for record_dir in holdout_dirs:
+      if os.path.exists(record_dir):  # make sure holdout dir exists
+        tf_records.extend(tf.gfile.Glob(os.path.join(record_dir, '*.zz')))
+
+  print('The length of tf_records is %d.' % len(tf_records))
+  with utils.logged_timer('Validating from %s to %s' % (
+      os.path.basename(tf_records[0]), os.path.basename(tf_records[-1]))):
+    dual_net.validate(estimator_working_dir, tf_records,
+                      checkpoint_name, params)
+
+
+def evaluate(models_dir, black_model_name, white_model_name,
+             evaluate_dir, params):
+  """Evaluate on two models.
+
+  Args:
+    models_dir: Directories where the completed generations/models are.
+    black_model_name: The name of the model playing black
+    white_model_name: The name of the model playing white
+    evaluate_dir: Where to write the evaluation results'='sgf/evaluate/
+    params: Hyperparameters.
+
+  Returns:
+    The model name of the winner.
+  """
+
+  black_model = os.path.join(models_dir, black_model_name)
+  white_model = os.path.join(models_dir, white_model_name)
+
+  print('Evaluate models between %s and %s' % (
+      black_model_name, white_model_name))
+
+  _ensure_dir_exists(evaluate_dir)
+
+  with utils.logged_timer('Loading weights'):
+    black_net = dual_net.DualNetwork(black_model, params)
+    white_net = dual_net.DualNetwork(white_model, params)
+
+  with utils.logged_timer('%d games' % params.eval_games):
+    winner = evaluation.play_match(
+        black_net, white_net, params.eval_games, params.eval_readouts,
+        evaluate_dir, params.eval_verbose)
+
+  return black_model_name if winner == 'B' else white_model_name
+
+
+def main(_):
+  """Run the reinforcement learning loop."""
+  tf.logging.set_verbosity(tf.logging.INFO)
+
+  params = model_utils.MiniGoModelParams()
+
+  k = utils.round_power_of_two(FLAGS.go_size ** 2 / 3)
+  params.num_filters = k  # Number of filters in the convolution layer
+  params.fc_width = 2 * k  # Width of each fully connected layer
+  params.num_shared_layers = FLAGS.go_size  # Number of shared trunk layers
+  params.go_size = FLAGS.go_size  # Board size
+
+  # set the shuffle buffer size smaller
+  preprocessing.SHUFFLE_BUFFER_SIZE = 200000
+
+  # How many positions can fit on a graphics card. 256 for 9s, 16 or 32 for 19s.
+  if FLAGS.go_size == 9:
+    params.batch_size = 256
```

Comment:
We should be able to set batch_size via command line, I think... because graphics cards can very greatly.